doc: Merges CLOUDP-274025 docs & examples branch - DONT MERGE YET (#3079)
mongodb/terraform-provider-mongodbatlas: docs/data-sources/advanced_cluster (preview provider v2).md

# Data Source: mongodbatlas_advanced_cluster (Preview for MongoDB Atlas Provider v2)

`mongodbatlas_advanced_cluster` describes an Advanced Cluster. The data source requires your Project ID.

> This doc is for the **Preview for MongoDB Atlas Provider v2** of `mongodbatlas_advanced_cluster`, doc for current version can be found [here](./advanced_cluster). In order to enable the Preview, you must set the enviroment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true`, otherwise the current version will be used.

Reviewer: ```suggestion
This page describes the **Preview for MongoDB Atlas Provider v2** of the `mongodbatlas_advanced_cluster` data source. For information about the current version, see [Data Source: mongodbatlas_advanced_cluster](./advanced_cluster). To enable the Preview, you must set the environment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true`; otherwise the current version will be used.
```

Reviewer: do we have a link or anything to an example of where a user would set this env variable?

Author: we have a section "Enable the..." in this [guide](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/CLOUDP-274025-dev-docs-examples/docs/guides/advanced-cluster-preview-provider-v2.md), although there is no much to say than setting the env var, as the way will depend if it's a local environment, Terraform Cloud, etc.

Reviewer: after looking through the other guides, I don't think we need to provide a link

> This doc is for the **Preview for MongoDB Atlas Provider v2** of `mongodbatlas_advanced_cluster`, doc for current version can be found [here](./advanced_cluster). In order to enable the Preview, you must set the enviroment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true`, otherwise the current version will be used.

Reviewer 2: ```suggestion
+ This doc is for the **Preview for MongoDB Atlas Provider v2** of `mongodbatlas_advanced_cluster`, and the documentation for the current version can be found [here](./advanced_cluster). In order to enable the Preview, you must set the enviroment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true`, otherwise the current version will be used.
```

Author: thanks, we're waiting for docs team review about using "doc", "info" as Marco flagged, will change after their review

-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

~> **IMPORTANT:**
<br> &#8226; Changes to cluster configurations can affect costs. Before making changes, please see [Billing](https://docs.atlas.mongodb.com/billing/).
<br> &#8226; If your Atlas project contains a custom role that uses actions introduced in a specific MongoDB version, you cannot create a cluster with a MongoDB version less than that version unless you delete the custom role.

## Example Usage

```terraform
resource "mongodbatlas_advanced_cluster" "example" {
  project_id   = "<YOUR-PROJECT-ID>"
  name         = "cluster-test"
  cluster_type = "REPLICASET"

  replication_specs = [{
    region_configs = [{
      electable_specs = {
        instance_size = "M0"
      }
      provider_name         = "TENANT"
      backing_provider_name = "AWS"
      region_name           = "US_EAST_1"
      priority              = 7
    }]
  }]
}

data "mongodbatlas_advanced_cluster" "example" {
	project_id = mongodbatlas_advanced_cluster.example.project_id
	name 	   = mongodbatlas_advanced_cluster.example.name
}
```

> **NOTE:** There can only be one M0 cluster per project.

Reviewer 3: do we need to add this? I don't think we should replicate Atlas limitations in our docs. I'd instead add a link to official Atlas documentation that mentions this.

Author: this is not parf of this PR, it's what is already in adv_cluster doc. But we can delete the line if you think it shouldn't be there, cc @oarbusi to confirm

Reviewer 3: yeah I don't think this should be here. These limits can easily be changed in the future so I'd just add atlas docs link

Reviewer 2: I think it makes sense to remove and link to atlas docs

Author: changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/397438f2c1e28859ee1cc3a2e7aea97fa66a8091

> **NOTE:** There can only be one M0 cluster per project.

Reviewer: is this still true? I know we've started rolling out a feature where select customers can use more than one M0, but not sure if it is configurable using terraform

Author: thanks, I already deleted this note as per another feedback

* `config_server_type` Describes a sharded cluster's config server type. Valid values are `DEDICATED` and `EMBEDDED`. To learn more, see the [Sharded Cluster Config Servers documentation](https://dochub.mongodb.org/docs/manual/core/sharded-cluster-config-servers/).

### bi_connector_config

Specifies BI Connector for Atlas configuration.

* `enabled` - Specifies whether or not BI Connector for Atlas is enabled on the cluster.l
* `read_preference` - Specifies the read preference to be used by BI Connector for Atlas on the cluster. Each BI Connector for Atlas read preference contains a distinct combination of [readPreference](https://docs.mongodb.com/manual/core/read-preference/) and [readPreferenceTags](https://docs.mongodb.com/manual/core/read-preference/#tag-sets) options. For details on BI Connector for Atlas read preferences, refer to the [BI Connector Read Preferences Table](https://docs.atlas.mongodb.com/tutorial/create-global-writes-cluster/#bic-read-preferences).

> ### Tags

Reviewer: should this be lower case to align with the field name
```suggestion
### tags
```

Author: thanks, changed all occurrences here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/b3d5da7df3d7a2593c429f715596883715bff4a2

* `value` - Variable that belongs to the set of the tag.

To learn more, see [Resource Tags](https://dochub.mongodb.org/core/add-cluster-tag-atlas).

### labels

Key-value pairs that categorize the cluster. Each key and value has a maximum length of 255 characters.  You cannot set the key `Infrastructure Tool`, it is used for internal purposes to track aggregate usage.

* `key` - The key that you want to write.
* `value` - The value that you want to write.

-> **NOTE:** MongoDB Atlas doesn't display your labels.

### replication_specs

* `id` - **(DEPRECATED)** Unique identifer of the replication document for a zone in a Global Cluster. This value corresponds to the legacy sharding schema (no independent shard scaling) and is different from the Shard ID you may see in the Atlas UI. This value is not populated (empty string) when a sharded cluster has independently scaled shards.
* `external_id` - Unique 24-hexadecimal digit string that identifies the replication object for a shard in a Cluster. This value corresponds to Shard ID displayed in the UI. When using old sharding configuration (replication spec with `num_shards` greater than 1) this value is not populated.

> * `num_shards` - Provide this value if you set a `cluster_type` of `SHARDED` or `GEOSHARDED`. **(DEPRECATED.)** To learn more, see the [Migration Guide](../guides/1.18.0-upgrade-guide.html.markdown).

Reviewer 3: ```suggestion
* `num_shards` - Provide this value if you set a `cluster_type` of `SHARDED` or `GEOSHARDED`. **(DEPRECATED)** To learn more, see the [Migration Guide](../guides/1.18.0-upgrade-guide.html.markdown).
```

Reviewer 3: would check in all places, there are a few occurences

Author: thanks, changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/cef152bdb5a120ec1f4179c70c0a22503d6526e6

### auto_scaling

* `disk_gb_enabled` - Flag that indicates whether this cluster enables disk auto-scaling. 
* `compute_enabled` - Flag that indicates whether instance size auto-scaling is enabled. 
* `compute_scale_down_enabled` - Flag that indicates whether the instance size may scale down. 
* `compute_min_instance_size` - Minimum instance size to which your cluster can automatically scale (such as M10). 
* `compute_max_instance_size` - Maximum instance size to which your cluster can automatically scale (such as M40). 

### analytics_auto_scaling

* `disk_gb_enabled` - Flag that indicates whether this cluster enables disk auto-scaling. 
* `compute_enabled` - Flag that indicates whether instance size auto-scaling is enabled. 
* `compute_scale_down_enabled` - Flag that indicates whether the instance size may scale down. 
* `compute_min_instance_size` - Minimum instance size to which your cluster can automatically scale (such as M10). 
* `compute_max_instance_size` - Maximum instance size to which your cluster can automatically scale (such as M40). 
#### Advanced Configuration

> * `default_read_concern` - [Default level of acknowledgment requested from MongoDB for read operations](https://docs.mongodb.com/manual/reference/read-concern/) set for this cluster. MongoDB 4.4 clusters default to [available](https://docs.mongodb.com/manual/reference/read-concern-available/). **(DEPRECATED.)** MongoDB 5.0 and later clusters default to `local`. To use a custom read concern level, please refer to your driver documentation.

Reviewer 3: ```suggestion
* `default_read_concern` - [Default level of acknowledgment requested from MongoDB for read operations](https://docs.mongodb.com/manual/reference/read-concern/) set for this cluster. MongoDB 4.4 clusters default to [available](https://docs.mongodb.com/manual/reference/read-concern-available/). **(DEPRECATED)** MongoDB 5.0 and later clusters default to `local`. To use a custom read concern level, please refer to your driver documentation.
```

Author: thanks, changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/cef152bdb5a120ec1f4179c70c0a22503d6526e6

* `default_write_concern` -  [Default level of acknowledgment requested from MongoDB for write operations](https://docs.mongodb.com/manual/reference/write-concern/) set for this cluster. MongoDB 4.4 clusters default to [1](https://docs.mongodb.com/manual/reference/write-concern/).

> * `fail_index_key_too_long` - When true, documents can only be updated or inserted if, for all indexed fields on the target collection, the corresponding index entries do not exceed 1024 bytes. When false, mongod writes documents that exceed the limit but does not index them. **(DEPRECATED.)** This parameter has been removed as of [MongoDB 4.4](https://www.mongodb.com/docs/manual/reference/parameters/#mongodb-parameter-param.failIndexKeyTooLong).

Reviewer 3: ```suggestion
* `fail_index_key_too_long` - When true, documents can only be updated or inserted if, for all indexed fields on the target collection, the corresponding index entries do not exceed 1024 bytes. When false, mongod writes documents that exceed the limit but does not index them. **(DEPRECATED)** This parameter has been removed as of [MongoDB 4.4](https://www.mongodb.com/docs/manual/reference/parameters/#mongodb-parameter-param.failIndexKeyTooLong).
```

Author: thanks, changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/cef152bdb5a120ec1f4179c70c0a22503d6526e6

      }
      provider_name = "AWS"
      priority      = 7
      region_name   = "EU_WEST_1"
    }]
    }, {
    region_configs = [{
      electable_specs = {
        instance_size = "M40"
        disk_iops     = 3000
        node_count    = 3
      }
      provider_name = "AWS"
      priority      = 7
      region_name   = "EU_WEST_1"
    }]
  }]
}

data "mongodbatlas_advanced_cluster" "example" {
  project_id                     = mongodbatlas_advanced_cluster.example.project_id
  name                           = mongodbatlas_advanced_cluster.example.name
  use_replication_spec_per_shard = true
}
```

## Example using Flex cluster

```terraform
resource "mongodbatlas_advanced_cluster" "example-flex" {
  project_id   = "<YOUR-PROJECT-ID>"
  name         = "flex-cluster"
  cluster_type = "REPLICASET"
  
  replication_specs {
    region_configs {
      provider_name = "FLEX"
      backing_provider_name = "AWS"
      region_name = "US_EAST_1"
      priority = 7
    }
  }
}
data "mongodbatlas_advanced_cluster" "example" {
  project_id = mongodbatlas_advanced_cluster.example-flex.project_id
  name       = mongodbatlas_advanced_cluster.example-flex.name
}
```

## Argument Reference

> * `project_id` - (Required) The unique ID for the project to create the database user.

Reviewer: ```suggestion
* `project_id` - (Required) The unique ID for the project to create the cluster.
```

Author: thanks, all occurrences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/e57eebd13be5d980358dce9ef0f3a0a0f719d7da

* `name` - (Required) Name of the cluster as it appears in Atlas. Once the cluster is created, its name cannot be changed.
* `use_replication_spec_per_shard` - (Optional) Set this field to true to allow the data source to use the latest schema representing each shard with an individual `replication_specs` object. This enables representing clusters with independent shard scaling.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` - The cluster ID.

> * `bi_connector_config` - Configuration settings applied to BI Connector for Atlas on this cluster. See [below](#bi_connector_config). **NOTE** Prior version of provider had parameter as `bi_connector`

Reviewer: ```suggestion
* `bi_connector_config` - Configuration settings applied to BI Connector for Atlas on this cluster. See [below](#bi_connector_config). **NOTE** In prior versions of the MongoDB Atlas Terraform Provider, this parameter was named `bi_connector`.
```

Author: thanks, all occurrences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/5030a6036d6b2c8976d836f8e09be9a1e92acfa3

* `replica_set_scaling_strategy` - (Optional) Replica set scaling mode for your cluster.
* `redact_client_log_data` - (Optional) Flag that enables or disables log redaction, see the [manual](https://www.mongodb.com/docs/manual/administration/monitoring/#log-redaction) for more info.
* `config_server_management_mode` - Config Server Management Mode for creating or updating a sharded cluster. Valid values are `ATLAS_MANAGED` (default) and `FIXED_TO_DEDICATED`. When configured as `ATLAS_MANAGED`, Atlas may automatically switch the cluster's config server type for optimal performance and savings. When configured as `FIXED_TO_DEDICATED`, the cluster will always use a dedicated config server. To learn more, see the [Sharded Cluster Config Servers documentation](https://dochub.mongodb.org/docs/manual/core/sharded-cluster-config-servers/).
* `config_server_type` Describes a sharded cluster's config server type. Valid values are `DEDICATED` and `EMBEDDED`. To learn more, see the [Sharded Cluster Config Servers documentation](https://dochub.mongodb.org/docs/manual/core/sharded-cluster-config-servers/).

### bi_connector_config

Specifies BI Connector for Atlas configuration.

> * `enabled` - Specifies whether or not BI Connector for Atlas is enabled on the cluster.l

Reviewer: ```suggestion
* `enabled` - Specifies whether or not BI Connector for Atlas is enabled on the cluster.
```

Author: thanks, changed all occurrences here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/a69a01e66154c70782e146529d97563ad8c1fc76

* `region_configs` - Configuration for the hardware specifications for nodes set for a given regionEach `region_configs` object describes the region's priority in elections and the number and type of MongoDB nodes that Atlas deploys to the region. Each `region_configs` object must have either an `analytics_specs` object, `electable_specs` object, or `read_only_specs` object. See [below](#region_configs)
*  `container_id` - A key-value map of the Network Peering Container ID(s) for the configuration specified in `region_configs`. The Container ID is the id of the container either created programmatically by the user before any clusters existed in a project or when the first cluster in the region (AWS/Azure) or project (GCP) was created.  The syntax is `"providerName:regionName" = "containerId"`. Example `AWS:US_EAST_1" = "61e0797dde08fb498ca11a71`.
* `zone_name` - Name for the zone in a Global Cluster.
* `zone_id` - Unique 24-hexadecimal digit string that identifies the zone in a Global Cluster. If clusterType is GEOSHARDED, this value indicates the zone that the given shard belongs to and can be used to configure Global Cluster backup policies.

### region_configs

> * `analytics_specs` - Hardware specifications for [analytics nodes](https://docs.atlas.mongodb.com/reference/faq/deployment/#std-label-analytics-nodes-overview) needed in the region. See [below](#specs)

Reviewer: ```suggestion
* `analytics_specs` - Hardware specifications for [analytics nodes](https://docs.atlas.mongodb.com/reference/faq/deployment/#std-label-analytics-nodes-overview) needed in the region. See [below](#specs).
```

Author: changes, all occurences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/a69236667047e801fd0b7e63cdeb29d86a0ed56d and here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/848601cc30d6b173a2d7ee741205a82961dd16ee

> * `auto_scaling` - Configuration for the Collection of settings that configures auto-scaling information for the cluster. See [below](#auto_scaling)

Reviewer: ```suggestion
* `auto_scaling` - Configuration for the Collection of settings that configures auto-scaling information for the cluster. See [below](#auto_scaling).
```

Author: changes, all occurences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/a69236667047e801fd0b7e63cdeb29d86a0ed56d and here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/848601cc30d6b173a2d7ee741205a82961dd16ee

> * `analytics_auto_scaling` - Configuration for the Collection of settings that configures analytics-auto-scaling information for the cluster. See [below](#analytics_auto_scaling)

Reviewer: ```suggestion
* `analytics_auto_scaling` - Configuration for the Collection of settings that configures analytics-auto-scaling information for the cluster. See [below](#analytics_auto_scaling).
```

Author: thanks, all occurences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/a69236667047e801fd0b7e63cdeb29d86a0ed56d and here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/848601cc30d6b173a2d7ee741205a82961dd16ee

### auto_scaling

* `disk_gb_enabled` - Flag that indicates whether this cluster enables disk auto-scaling. 
* `compute_enabled` - Flag that indicates whether instance size auto-scaling is enabled. 
* `compute_scale_down_enabled` - Flag that indicates whether the instance size may scale down. 
* `compute_min_instance_size` - Minimum instance size to which your cluster can automatically scale (such as M10). 
* `compute_max_instance_size` - Maximum instance size to which your cluster can automatically scale (such as M40). 

### analytics_auto_scaling

* `disk_gb_enabled` - Flag that indicates whether this cluster enables disk auto-scaling. 
* `compute_enabled` - Flag that indicates whether instance size auto-scaling is enabled. 
* `compute_scale_down_enabled` - Flag that indicates whether the instance size may scale down. 
* `compute_min_instance_size` - Minimum instance size to which your cluster can automatically scale (such as M10). 
* `compute_max_instance_size` - Maximum instance size to which your cluster can automatically scale (such as M40). 
#### Advanced Configuration

> * `default_read_concern` - [Default level of acknowledgment requested from MongoDB for read operations](https://docs.mongodb.com/manual/reference/read-concern/) set for this cluster. MongoDB 4.4 clusters default to [available](https://docs.mongodb.com/manual/reference/read-concern-available/). **(DEPRECATED)** MongoDB 5.0 and later clusters default to `local`. To use a custom read concern level, please refer to your driver documentation.

Reviewer: 4.4 is EOL. Can we remove this sentence? 5.0 is EOL, too, so consider changing the 5.0 reference to 6.0

Author: thanks, as this is not specific to the v2 changes, I've created a ticket to follow this: CLOUDP-302166

Author: done in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3220

> * `default_write_concern` -  [Default level of acknowledgment requested from MongoDB for write operations](https://docs.mongodb.com/manual/reference/write-concern/) set for this cluster. MongoDB 4.4 clusters default to [1](https://docs.mongodb.com/manual/reference/write-concern/).

Reviewer: what is the default for 6.0?

Author: thanks, as this is not specific to the v2 changes, I've created a ticket to follow this: CLOUDP-302166

Author: changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3220

> * `fail_index_key_too_long` - When true, documents can only be updated or inserted if, for all indexed fields on the target collection, the corresponding index entries do not exceed 1024 bytes. When false, mongod writes documents that exceed the limit but does not index them. **(DEPRECATED)** This parameter has been removed as of [MongoDB 4.4](https://www.mongodb.com/docs/manual/reference/parameters/#mongodb-parameter-param.failIndexKeyTooLong).

Reviewer: should this be removed given 4.4 is EOL?

Author: thanks, as this is not specific to the v2 changes, I've created a ticket to follow this: CLOUDP-302166

Author: changed in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3220

* `javascript_enabled` - When true, the cluster allows execution of operations that perform server-side executions of JavaScript. When false, the cluster disables execution of those operations.

> * `minimum_enabled_tls_protocol` - Sets the minimum Transport Layer Security (TLS) version the cluster accepts for incoming connections.Valid values are:

Reviewer: ```suggestion
* `minimum_enabled_tls_protocol` - Sets the minimum Transport Layer Security (TLS) version the cluster accepts for incoming connections. Valid values are:
```

Author: thanks, all occurrences changed here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3079/commits/b19ca55ddefd77ebdaeaa6ffba3c7275e460d043

  - TLS1_0
  - TLS1_1

>   - TLS1_2

Reviewer: q: does TLS1_3 need to be added?

Reviewer: might need to update this upstream as well

Author: thanks, as this is not specific to the v2 changes, I've created a ticket to follow this: CLOUDP-302166

Author: added here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3220
